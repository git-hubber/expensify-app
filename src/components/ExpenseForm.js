import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
      buttonText: props.expense ? 'Edit Expense' : 'Add Expense',
    };
  }

  _onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  _onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  _onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  _onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  _onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  _onFormSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please be sure to provide the description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));

      const expense = {
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10), // * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      };
      this.props.onSubmit(expense);
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this._onFormSubmit}>
        {this.state.error && <p style={{ color: 'red' }}>{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={this.state.description}
          onChange={this._onDescriptionChange}
        />
        <input
          type="text"
          placeholder="amount"
          className="text-input"
          value={this.state.amount}
          onChange={this._onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this._onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this._onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="textarea"
          value={this.state.note}
          onChange={this._onNoteChange}
        />
        <div>
          <button className="button">{this.state.buttonText}</button>
        </div>
      </form>
    );
  }
}

export default ExpenseForm;
