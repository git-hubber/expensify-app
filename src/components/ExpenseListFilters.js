import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }

  _onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  _onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }

  _onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  _onSortChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this._onTextChange}
        />
        <select
          onChange={this._onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this._onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this._onFocusChange}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);