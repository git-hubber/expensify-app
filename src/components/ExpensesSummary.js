import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
  const totalExpenses = props.expenses.length;
  const totalAmount = numeral(selectExpensesTotal(props.expenses))
    .format('$0,0.00');
  const word = totalExpenses === 1 ? 'expense' : 'expenses';

  return (
    <div>
      Viewing {totalExpenses} {word} totalling {totalAmount}
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpensesSummary);
