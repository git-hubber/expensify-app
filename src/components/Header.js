import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';

import { startLogout } from '../actions/auth'

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" exact={true} activeClassName="is-active">Home</NavLink> -
    <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    <button onClick={ startLogout }>LogOut</button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = ( { auth }) => {
  console.log('Auth:' , auth);
  return auth;
}

export default connect(mapStateToProps, mapDispatchToProps)(Header);