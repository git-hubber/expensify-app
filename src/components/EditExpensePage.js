import React, { Component } from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
  _onSubmit = (expense) => {
    this.props.editExpense(expense.id, expense);
    this.props.history.push("/");
  };

  _onRemove = () => {
    this.props.startRemoveExpense( { id: this.props.expense.id } );
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this._onSubmit}
        />
        <button onClick={this._onRemove}>
          Remove
        </button>
      </div>
    )
  }
}

const mapStateToProps = ({ expenses }, props) => ({
  expense: expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
