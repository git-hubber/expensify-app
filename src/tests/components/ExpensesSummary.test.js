import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should correctly render with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render with 2 expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[0], expenses[1]]} />);
  expect(wrapper).toMatchSnapshot();
});
