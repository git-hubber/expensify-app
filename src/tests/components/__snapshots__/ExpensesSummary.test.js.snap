// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $195.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render with 2 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $109,695.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
